### PR TITLE
Change default --max-pods in kubelet to 110

### PR DESCRIPTION
**Pull request by yujuhong**

### CODE REVIEW

**Comment by yujuhong** [comment]
ref: #16943 

/cc @kubernetes/sig-scalability @kubernetes/sig-node

**Comment by dchen1107** [comment]
Sorry, just saw you are also working on this one. 

I have a similar one #21365, which also includes max-dead-containers change. We need this change when we bump up max-pods too.

File: cmd/kubelet/app/options/options.go [hunk 1]
```diff
@@ -94,7 +94,7 @@ func NewKubeletServer() *KubeletServer {
 			MaxContainerCount:           100,
```
Comment (feiskyer) [inline cmd/kubelet/app/options/options.go]: Should `MaxConainerCount` increased to 200?

Comment (feiskyer) [inline cmd/kubelet/app/options/options.go]: Should `MaxConainerCount` be increased to 200?

**Comment by timothysc** [comment]
Could you update the defaults in the e2e tests too otherwise lgtm.

**Comment by kelseyhightower** [comment]
Is there a change log or release notes that get updated with changes like this?

**Comment by wojtek-t** [comment]
Yes - I think this should be added to release notes - we should probably add a label "release-note"

**Comment by jeremyeder** [comment]
We haven't finished testing the most recent rebase of Kube into OpenShift (so the OSE data below will improve further), but I wanted to LGTM this PR as well.

Single node, 120-sleep pod test shows great efficiency improvements for node vertical scale.

- Reduction from ~3 cores (just for the infrastructure pids) down to below 0.5 cores.
cpu

- Reduction from 1100MB down to 250MB memory.
ram

File: cmd/kubelet/app/options/options.go (hunk 1 above)
Comment (yujuhong) [inline cmd/kubelet/app/options/options.go]: Yes, we should probably increase this. Unfortunately, our container GC implementation is far from perfect. This literally means *when* the GC will kick in. By changing this to 200, the GC won't start until there are 200 dead containers, even if you only run 3 pods on the node. /cc @dchen1107 @vishh

**Comment by yujuhong** [comment]
Let's get this PR in first. I'll follow up with the test changes and garbage collection issues in #21570

**Comment by wojtek-t** [comment]
>
Let's get this PR in first. I'll follow up with the test changes and garbage collection issues in #21570 

SGTM.

Also - I'm adding release-note label.

**Comment by gmarek** [comment]
Don't you want to set it to 105 or sth, to account for system pods?

**Comment by yujuhong** [comment]
> Don't you want to set it to 105 or sth, to account for system pods?

@dchen1107, can we do that and still only advertise 100 pods per node? If so, I am okay with 105.

The alternatives:
 - Set max pods to 100 and then test ~95 pause pods per node (to account for the system pods)
 - Set max pods to 105 in an e2e test cluster.

**Comment by yujuhong** [comment]
> @dchen1107, can we do that and still only advertise 100 pods per node? If so, I am okay with 105.

I decided to go with 110 to make testing gke clusters easier (which have no separate test configurations). 
We should still advertise 100 user pods.

**Comment by ixdy** [comment]
@k8s-bot unit test this again please, issue #IGNORE (I accidentally broke one of the Jenkins slaves)

**Comment by ixdy** [comment]
@k8s-bot unit test this again please, issue #IGNORE (I accidentally broke one of the Jenkins slaves)

(sorry again)

**Comment by yujuhong** [comment]
@k8s-bot test this please, issue #IGNORE

**Comment by dchen1107** [comment]
LGTM

**Comment by wojtek-t** [comment]
LGTM

**Comment by jeremyeder** [comment]
@yujuhong @dchen1107 @timothysc 

```@dchen1107, can we do that and still only advertise 100 pods per node? If so, I am okay with 105.
I decided to go with 110 to make testing gke clusters easier (which have no separate test configurations). 
We should still advertise 100 user pods.
```

What is the difference between the PR code and "advertise".  Is advertise what's documented?  I can't see how you can differentiate in docs but have them different than the defaults.  That means any cluster that users install would have max-pods greater than what is actually "advertised/supported".  Unless "advertise" means something special?

Otherwise certainly OpenShift will have to carry a patch to our kubelet configuration that reduces max-pods to what is "advertised" :-1: 

My thought is that the tests themselves should be fixed to take a runtime argument for the kubelet.  Otherwise you will face this same strangeness every time you want to bump the defaults.

Also in the PR you have max-containers set to 240.  That should be either 220 or 200 based on max-pods, shouldn't it?

I guess what I'm saying is that the 110 has a strange reasoning behind it.  Any chance you can just go with 100?

**Comment by timstclair** [comment]
I agree with @jeremyeder , the 110 number will be shown if someone queries the kubelet flags (and eventually documented in the config API), and it looks a little arbitrary. If we have wiggle room for system pods anyway, what if we just don't count system pods towards the maximum?

**Comment by yujuhong** [comment]
> I guess what I'm saying is that the 110 has a strange reasoning behind it. Any chance you can just go with 100?

I agree it seems strange, that's why initially I set it to 100. However, this seems to be the convention. For v1.0, we set MaxPods to 40, but documented that we support 30 pods per node. (https://github.com/kubernetes/kubernetes/blob/3364a32dc6ad8559724249fa5dc9e549cee00139/docs/admin/cluster-large.md)
I believe that it took into account that we have cluster addon pods that would take up resources, i.e., technically kubelet supports 40 pods per node, but that translates into ~30 user pods.


> My thought is that the tests themselves should be fixed to take a runtime argument for the kubelet. Otherwise you will face this same strangeness every time you want to bump the defaults.

We do overwrite this parameter in a GCE test cluster, but is not yet able to do it for a GKE cluster. We can always create 95 extra pods per node in a test to work around it.


> Also in the PR you have max-containers set to 240. That should be either 220 or 200 based on max-pods, shouldn't it?

The max-containers is just a trigger to forcibly delete all old, dead containers given today's implementation. It doesn't have to be MaxPods*2. I left a buffer following the same pattern it was.
We have another issue open to fix the GC behavior for v1.2, which will render this global cap less meaningful. 


I started the PR with 100 and changed it to 110, so I see the pros and cons. There are also discussion in @dchen1107's PR: https://github.com/kubernetes/kubernetes/pull/21365#issuecomment-187431772
I am fine with both and I'll let @dchen1107 make the final decision. Removing the lgtm label so we'll have time to settle this issue.

**Comment by vishh** [comment]
If we can support `110` in reality, we should set that as the limit. Addons are provider specific. OSS kube should mention the limit as `110` (if we choose that to be the limit). GKE (and other providers), can then advertise a number lesser than that based on the addons that are being run in the cluster.

**Comment by gmarek** [comment]
We always thought of it as 'max-user-pods' thing.

**Comment by dchen1107** [comment]
Yes, we made the decision on configuring max-pods to 40 for 1.0 release because we announce 30 user pods per node, there 10 pods for kube-system. This change is align with that decision.

**Comment by dchen1107** [comment]
LGTM

**Comment by wojtek-t** [comment]
I agree with @dchen1107 - LGTM

**Comment by wojtek-t** [comment]
@k8s-bot e2e test this please github issue: #21261

**Pull request merged**